src/hotspot/share/runtime/atomic.hpp

```diff
@@ -1232,4 +1251,492 @@ inline T Atomic::XchgUsingCmpxchg<byte_size>::operator()(T volatile* dest,
+  using Base::release_store_fence;
+  using Base::cmpxchg;
+  using Base::xchg;
+  using Base::add_then_fetch;
+  using Base::fetch_then_add;
+  using Base::sub_then_fetch;
+  using Base::fetch_then_sub;
+  using Base::atomic_inc;
+  using Base::atomic_dec;
+
+  bool replace_if_null(T exchange_value,
+                       atomic_memory_order order = memory_order_conservative) {
+    return Atomic::replace_if_null(value_ptr(), exchange_value, order);
+  }
+
+  bool clear_if(T compare_value,
+                atomic_memory_order order = memory_order_conservative) {
+    return compare_value == cmpxchg(compare_value, (T)nullptr, order);
+  }
+
+  T fetch_then_clear(atomic_memory_order order = memory_order_conservative) {
+    return xchg((T)nullptr);
+  }
+};
+
+template<typename T>
+class Atomic::AtomicValueHasExchange {
+  template<typename Check> static char* test(decltype(&Check::xchg));
+  template<typename> static char test(...);
+  using test_type = decltype(test<AtomicValue<T>>(nullptr));
+public:
+  static const bool value = std::is_pointer_v<test_type>;
+};
+
+template<typename T>
+class Atomic::TranslatingValue {
+  using Translator = PrimitiveConversions::Translate<T>;
+  using Decayed = typename Translator::Decayed;
+
+  AtomicValue<Decayed> _value;
+
+  static Decayed decay(T x) { return Translator::decay(x); }
+  static T recover(Decayed x) { return Translator::recover(x); }
+
+  // "Unit test" AtomicValueHasExchange.
+  static_assert(AtomicValueHasExchange<int>::value);
+  static_assert(AtomicValueHasExchange<void*>::value);
+  static_assert(!AtomicValueHasExchange<char>::value);
+
+public:
+  using ValueType = T;
+
+  template<typename Dummy = int,
+           std::enable_if_t<std::is_default_constructible_v<T>, Dummy> = 0>
+  TranslatingValue() : TranslatingValue(T()) {}
+
+  explicit TranslatingValue(T value) : _value(decay(value)) {}
+
+  NONCOPYABLE(TranslatingValue);
+
+  static int value_offset_in_bytes() {
+    return (offset_of(TranslatingValue, _value) +
+            AtomicValue<Decayed>::value_offset_in_bytes());
+  }
+
+  static int value_size_in_bytes() {
+    return AtomicValue<Decayed>::value_size_in_bytes();
+  }
+
+  T load_relaxed() const {
+    return recover(_value.load_relaxed());
+  }
+
+  T load_acquire() const {
+    return recover(_value.load_acquire());
+  }
+
+  void relaxed_store(T value) {
+    _value.relaxed_store(decay(value));
+  }
+
+  void release_store(T value) {
+    _value.release_store(decay(value));
+  }
+
+  void release_store_fence(T value) {
+    _value.release_store_fence(decay(value));
+  }
+
+  T cmpxchg(T compare_value, T exchange_value,
+            atomic_memory_order order = memory_order_conservative) {
+    return recover(_value.cmpxchg(decay(compare_value),
+                                  decay(exchange_value),
+                                  order));
+  }
+
+  template<typename Dummy = int,
+           std::enable_if_t<AtomicValueHasExchange<Decayed>::value, Dummy> = 0>
```

Comment:
I have a macro elsewhere to package this little dance for non-dependent
SFINAE, a la `ENABLE_IF` but dealing with the lack of a dependency. The comment
describing `ENABLE_IF` talks about this "uncommon" case and not bothering to
provide a macro to support it, but I've run into it several times recently.
If we could rely on the availability of the `__COUNTER__` extension, `ENABLE_IF`
could just deal with it.